src/scheduler.cc

```diff
@@ -222,10 +232,19 @@ Status SchedulerService::SchedulerInfo(ServerContext* context, const SchedulerIn
 void SchedulerService::deliver_object(ObjRef objref, ObjStoreId from, ObjStoreId to) {
   RAY_CHECK_NEQ(from, to, "attempting to deliver objref " << objref << " from objstore " << from << " to itself.");
   RAY_CHECK(has_canonical_objref(objref), "attempting to deliver objref " << objref << ", but this objref does not yet have a canonical objref.");
+  ObjRef canonical_objref = get_canonical_objref(objref);
+  {
+    // We increment once so the objref doesn't go out of scope before the ObjReady
+    // method is called. The corresponding decrement will happen in ObjReady in
+    // the scheduler.
+    std::lock_guard<std::mutex> reference_counts_lock(reference_counts_lock_); // we grab this lock because increment_ref_count assumes it has been acquired
+    std::vector<ObjRef> objref_vector;
+    objref_vector.push_back(canonical_objref);
+    increment_ref_count(objref_vector);
+  }
```

Comment:
some of these are not valid because of the order in which the locks are acquired.